website/static/js/apiApplication.js

```diff
@@ -0,0 +1,317 @@
+    var validated = ko.validatedObservable(self);
+    self.isValid = ko.computed(function(){
+        return validated.isValid();
+    });
+
+};
+
+// Serialize data for POST request
+ApplicationData.prototype.toJSON = function () {
+    var self = this;
+    return {
+        client_id: self.clientId,
+        client_secret: self.clientSecret(),
+        owner: self.owner(),
+        name: self.name(),
+        description: self.description(),
+        home_url: self.homeUrl(),
+        callback_url: self.callbackUrl()
+    };
+    };
+
+// Load data from JSON
+ApplicationData.prototype.fromJSON = function (data) {
+    var self = this;
+    data = data || {};
+    self.clientId = data.client_id;
+    self.clientSecret(data.client_secret);
+    self.create_date = data.create_date;
+
+    self.owner(data.owner);
+    self.name(data.name);
+    self.description(data.description);
+    self.homeUrl(data.home_url);
+    self.callbackUrl(data.callback_url);
+
+    self.detailUrl = data.links.html;
+    self.apiUrl = data.links.self;
+};
+
+
+var applicationFetch = function (url) { // Function shared by list and detail ViewModels
+    var self = this;
+    var request = $osf.ajaxJSON('GET', url, {isCors: true});
+
+    request.done(function (data) {
+        var result;
+        // Check return type to handle both list and detail views
+        if (Array.isArray(data.data)) {  // ES5 dependent
+            result = $.map(data.data, function (item) {
+                return new ApplicationData(item);
+            });
+        } else if (data.data) {
+            result = new ApplicationData(data.data);
+        }
+        self.content(result);
+    });
+
+    request.fail(function (xhr, status, err) {
+        $osf.growl('Error', 'Data not loaded. Please refresh the page and try ' +
+            'again or contact <a href="mailto: support@cos.io">support@cos.io</a> ' +
+            'if the problem persists.', 'danger');
+
+        Raven.captureMessage('Error fetching application data', {
+            url: url,
+            status: status,
+            error: err
+        });
+    });
+};
+
+
+/*
+ * Create (or update) the data for a single application. On creation, page seamlessly transitions to acting
+ * like an update page.
+ */
+var ApplicationViewModel = function (urls) {
+    // Read and update operations
+
+    var self = this;
+    self.content = ko.observable(new ApplicationData());
+    self.message = ko.observable();
+    self.messageClass = ko.observable();
+
+    self.showMessages = ko.observable(false);
+
+    self.dataUrl = urls.dataUrl; // Use for editing existing instances
+    self.submitUrl = urls.submitUrl; // Use for creating new instances
+    self.listPageUrl = urls.listPageUrl;
+
+    if (self.dataUrl){ // Support detail views
+        self.fetch(self.dataUrl);
+    }
+};
+
+ApplicationViewModel.prototype.fetch = applicationFetch;
+
+ApplicationViewModel.prototype.updateApplication = function () {
```

Comment:
We are moving away from having too much HTTP work done in our ViewModels, as it's a poor separation of concerns.

A better pattern is to have the following entities:

- Model objects to store app data and any computeds
- A Client object that makes HTTP requests and returns Model
- A thin ViewModel which dispatches to the Client and updates the View accordingly

In theory, the Model and Client are reusable and framework-agnostic. In practice, the Model will contain `ko.observables` and `ko.computeds` so that we can easily integrate them with Knockout views. Even so, the pattern makes for better reusability and easier refactoring.

For a good example of this pattern, see https://github.com/CenterForOpenScience/osf.io/blob/b9495e80a8f68549b3f2683d0120b948645f4a1a/website/static/js/accountSettings.js